### DISCUSSION TITLE
nack requeue error

### DISCUSSION THREAD

**Discussion by ryan053857** [roadrunner-server/roadrunner#2056] [Show and tell]
In roadrunner-php jobs sdk, I found that the parameters of nack message passing are:

```php
$this->respond(Type::NACK, [
    'message' => (string) $message,
    'redelivery' => $redelivery,
    'delay_seconds' => $this->delay,
]);
```

and 

```php
private function respond(int $type, array $data = []): void
{
    if ($this->completed === null) {
        try {
            $body = \json_encode(['type' => $type, 'data' => $data], JSON_THROW_ON_ERROR);

            $this->worker->respond(new Payload($body));
        } catch (\JsonException $e) {
            throw new SerializationException($e->getMessage(), $e->getCode(), $e);
        } catch (\Throwable $e) {
            throw new JobsException($e->getMessage(), (int)$e->getCode(), $e);
        }
        $this->completed = $type;
    }
}
```

The message structure defined in golang's jobs plugin is:

```go
package protocol

type errorResp struct {
	Msg     string              `json:"message"`
	Delay   int                 `json:"delay_seconds"`
	Headers map[string][]string `json:"headers"`
	Requeue bool                `json:"requeue"`
}

```

One is `redelivery` and the other is `requeue`, which makes nack's requeue ineffective

**Comment by rustatian** [comment]
Hey @shellphy 👋 
Requeue is used on the RR's part. `respond` was preserved to not break backward compatibility. You should simply use nack.

**Reply by ryan053857** [reply]
I am using nats jobs, and nats jobs cannot use delay_seconds when requeue is used, but delay_seconds can be used in nack.

**Reply by rustatian** [reply]
Could you please rephrase? I'm not sure if I'm following.

**Reply by ryan053857** [reply]
That is, I am using the nats jobs driver and I want to delay retry if the task fails.

If I use `requeue()` with delay_seconds, the nats jobs driver is not supported:

```golang
func (c *Driver) requeue(item *Item) error {
	const op = errors.Op("nats_requeue")
	if item.Options.Delay > 0 {
		return errors.E(op, errors.Str("nats doesn't support delayed messages, see: https://github.com/nats-io/nats-streaming-server/issues/324"))
	}
    ...
...
```

If I use `nack()` with delay_seconds, the nats jobs driver is supported, and then the delay option will only take effect if `requeue=true`:

```golang
func (i *Item) NackWithOptions(requeue bool, delay int) error {
	if atomic.LoadUint64(i.Options.stopped) == 1 {
		return errors.Str("failed to NACK the JOB, the pipeline is probably stopped")
	}

	// if the user requested to requeue the message
	if requeue {
		return i.Options.nakWithDelay(time.Second * time.Duration(delay))
	}

	// if the user requested to delete the message
	return i.Options.term()
}
```

However, when passing `nack` messages in php, Only `redelevery` but no `requeue`

**Reply by rustatian** [reply]
Yes, this is how it works. `requeue` as you may see - private. You should use NackWithOptions.

**Reply by ryan053857** [reply]
But the `requeue` parameter is always false in `NackWithOptions`

**Comment by ryan053857** [comment]
Is there a plan to solve this problem?

**Reply by rustatian** [reply]
There are no problems to resolve:
<img width="849" alt="image" src="https://github.com/user-attachments/assets/1cdfdda9-3298-40bd-8341-2c20a3ce1b63">
In PHP, you use `NACK` + delay + redelivery=true, in Go it'll be `NackWithOptions` under the hood.

**Reply by ryan053857** [reply]
But in Go, the implementation is like this:

```go
func (i *Item) NackWithOptions(requeue bool, delay int) error {
	if atomic.LoadUint64(i.Options.stopped) == 1 {
		return errors.Str("failed to NACK the JOB, the pipeline is probably stopped")
	}

	// if the user requested to requeue the message
	if requeue {
		return i.Options.nakWithDelay(time.Second * time.Duration(delay))
	}

	// if the user requested to delete the message
	return i.Options.term()
}
```

`requeue` is always the `false`

**Reply by rustatian** [reply]
No, requeue is not always false.

**Reply by rustatian** [reply]
I'm not sure how did you determine, that requeue is always false. It is an in arg in the method. You can't say it's always false without the caller.

**Reply by ryan053857** [reply]
Because the response struct is:

```go
package protocol

type errorResp struct {
	Msg     string              `json:"message"`
	Delay   int                 `json:"delay_seconds"`
	Headers map[string][]string `json:"headers"`
	Requeue bool                `json:"requeue"`
}
```



No `requeue` field  is sent in `nack()`.

**Reply by rustatian** [reply]
If you're sure this is the bug - send a PR into the PHP library.

**Reply by ryan053857** [reply]
ok
